### PR TITLE
fix: correctly track sources for open LSP documents

### LINKED ISSUE #5562: Bug: LSP features don't work well with unsaved documents
### Aim

LSP features should work well if a file has unsaved contents. For example, document symbols, inlay hints, references, etc., should all work well if, say, a struct is moved to a difference place in a file but that file isn't yet saved: references to that struct should take you to the new location, even though the file is unsaved.

### Expected Behavior

This is expected behavior:

https://github.com/user-attachments/assets/65d4684e-16ca-4cff-a643-c8474144feaf

Note how when we edit the file but don't save it, inlay hints are correctly updated. We can even define a struct, add a reference to it and jump to it, all while the file is unsaved.

### Bug

This is how it currently works:

https://github.com/user-attachments/assets/e6d5a8b4-69ab-4884-980c-8fd6dcdd46ba

Note how inlay hints still are always based on the saved version of the file, and how you can't jump to "Foo".

### To Reproduce

_No response_

### Project Impact

None

### Impact Context

_No response_

### Workaround

None

### Workaround Description

_No response_

### Additional Context

_No response_

### Installation Method

None

### Nargo Version

_No response_

### NoirJS Version

_No response_

### Would you like to submit a PR for this Issue?

Yes

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by asterite**
# Description

## Problem

Resolves #5562

## Summary

I noticed that our LSP code puts unsaved files into an `input_files` dictionary, but that dictionary was not used when adding files to a FileManager, meaning that file sources always corresponded to disk contents, never unsaved content.

Additionally, when we got a notification that a file changed, only that file (without everything it depended on) was processed, just to get code lenses. That means that references to types and variables weren't updated here, so things like hover, go to reference and inlay hints wouldn't work well.

To fix this, a few things have been done in this PR:
1. When creating a FileManager, always add these unsaved files first.
2. When we get the notification that a file was changed, type-check only the current package but without producing errors (at least Rust Analyzer doesn't update the Problems list until you save a file). This is also how it used to work before this PR.
3. When a file is closed, we also type-check the current package to update references (to avoid keeping references to that unsaved code locations).

Type-checking a package on every edit sounds a bit too much... but I checked a few big crates and type-checking just a single package (in a workspace of many packages) takes around 300ms on my machine (with a debug build), so I think it's acceptable. If it's slower then it might still be good enough (it's worse if things break and you get incorrect inlay hints or error notifications). Maybe ideally we'd process all top-level definitions in a package, but only fully type-check (looking inside function definitions) in the file that just changed, but I couldn't find an easy way to do that, so that would be an optimization for later (if really needed).

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: tooling/lsp/src/notifications/mod.rs [hunk 1]
```diff
@@ -105,7 +75,12 @@ pub(super) fn on_did_close_text_document(
         state.cached_definitions.clear();
     }
 
-    ControlFlow::Continue(())
+    let document_uri = params.text_document.uri;
+
+    match process_workspace_for_noir_document(state, document_uri, true, false) {
```
Comment (michaeljklein) [inline tooling/lsp/src/notifications/mod.rs:80]: `let` bindings for multiple `bool`'s can be easier to read:
```suggestion
    let only_process_document_uri_package = true;
    let no_diagnostics = false;
    match process_workspace_for_noir_document(state, document_uri, only_process_document_uri_package, no_diagnostics) {
```

**Review by michaeljklein — CHANGES_REQUESTED** [review]

File: tooling/lsp/src/notifications/mod.rs [hunk 2]
```diff
@@ -176,46 +161,53 @@ pub(crate) fn process_workspace_for_noir_document(
...
+                    .filter_map(|FileDiagnostic { file_id, diagnostic, call_stack: _ }| {
+                        // Ignore diagnostics for any file that wasn't the file we saved
+                        // TODO: In the future, we could create "related" diagnostics for these files
+                        if fm.path(file_id).expect("file must exist to have emitted diagnostic")
+                            != file_path
+                        {
+                            return None;
+                        }
+
+                        // TODO: Should this be the first item in secondaries? Should we bail when we find a range?
```
[the remark this answers is not in the document]
Comment (asterite) [reply tooling/lsp/src/notifications/mod.rs:176]: Oh, this line was indented so it shows up in the diff... so I'm not sure what this TODO is about 😅

File: tooling/lsp/src/notifications/mod.rs [hunk 3]
```diff
@@ -176,46 +161,53 @@ pub(crate) fn process_workspace_for_noir_document(
...
-                        range,
-                        severity: Some(severity),
-                        message: diagnostic.message,
-                        ..Default::default()
+            if output_diagnostics {
+                file_diagnostics
+                    .into_iter()
+                    .filter_map(|FileDiagnostic { file_id, diagnostic, call_stack: _ }| {
+                        // Ignore diagnostics for any file that wasn't the file we saved
+                        // TODO: In the future, we could create "related" diagnostics for these files
```
[the remark this answers is not in the document]
Comment (asterite) [reply tooling/lsp/src/notifications/mod.rs:169]: Same, this was already like this, so I'm not sure what would be needed here.

File: tooling/lsp/src/notifications/mod.rs (hunk 1 above)
Comment (asterite) [reply tooling/lsp/src/notifications/mod.rs]: Makes sense. Done!

File: tooling/lsp/src/notifications/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (michaeljklein) [reply tooling/lsp/src/notifications/mod.rs:197]: Ah, I see. All good then 👍

**Review by michaeljklein — APPROVED** [review]
LGTM

**Comment by asterite** [comment]
@michaeljklein Thank you!

**Pull request merged**